RelationalAI/TransactionAsyncAbortReasons.cs

```diff
@@ -0,0 +1,54 @@
+﻿/*
+ * Copyright 2022 RelationalAI, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+using System.Runtime.Serialization;
+
+namespace RelationalAI
+{
+    public enum TransactionAsyncAbortReason
+    {
+        [EnumMember(Value = "system internal error")]
+        SystemInternalError,
+        [EnumMember(Value = "cancelled")]
+        Cancelled,
+        [EnumMember(Value = "engine shutting down")]
+        EngineShuttingDown,
+        [EnumMember(Value = "invalid request")]
+        InvalidRequest,
+        [EnumMember(Value = "too many requests")]
+        TooManyRequests,
+        [EnumMember(Value = "integrity constraint violation")]
+        IntegrityConstraintViolation
+    }
+
+    public static class TransactionAsyncAbortReasons
+    {
+        public static string Value(this TransactionAsyncAbortReason reason)
+        {
+            return reason switch
+            {
+                TransactionAsyncAbortReason.SystemInternalError => "system internal error",
+                TransactionAsyncAbortReason.Cancelled => "cancelled",
+                TransactionAsyncAbortReason.EngineShuttingDown => "engine shutting down",
+                TransactionAsyncAbortReason.InvalidRequest => "invalid request",
+                TransactionAsyncAbortReason.TooManyRequests => "too many requests",
+                TransactionAsyncAbortReason.IntegrityConstraintViolation => "integrity constraint violation",
+                _ => throw new ArgumentOutOfRangeException(nameof(reason), reason, "unknow async transaction abort reason")
+            };
+        }
+    }
```

Comment:
I think we had previously decided that enums are dangerous in the SDK, since if we add another abort reason, it will break old SDKs (they will throw this "unknown async abort reason" exception).

I think it's better to just define the strings as _constants_, and then check for the expected string constant.

It does feel a bit scary to depend on string constants: what if we later rename `"integrity constraint violation"` to `"ic violation"` or something, then things would break!
And that's true. But I think it's better to have the old SDKs gracefully degrade, rather than throw an exception. Does that make sense?

So in this case, if we hardcode things to match the string `"integrity constraint violation"`, and only fetch the metadata in that case, then even if the string changes in a future release, it simply means that we will stop fetching metadata, and the user won't see their violations, but at least things won't crash?

But still, I think we should just be careful **not to change the strings**, since the SDKs rely on them. 👍 

The main issue with an enum is that _adding a new abort reason would break things,_ and that's not desirable.

Does that make sense?

Reply:
thanks @NHDaly addressed your comments can you give the PR another check :pray: ?